src/recovery.rs

```diff
@@ -341,3 +341,50 @@ impl<I: Clone, O, A: Parser<I, O, Error = E>, S: Strategy<I, O, E>, E: Error<I>>
         self.parse_inner(d, s)
     }
 }
+
+/// A parser that includes a fallback recovery strategy should parsing result in an error.
+#[derive(Copy, Clone)]
+pub struct RecoverVia<A, S>(pub(crate) A, pub(crate) S);
+
+impl<I: Clone, O, A: Parser<I, O, Error = E>, B: Parser<I, O, Error = E>, E: Error<I>> Parser<I, O>
+    for RecoverVia<A, B>
+{
+    type Error = E;
+
+    fn parse_inner<D: Debugger>(
+        &self,
+        debugger: &mut D,
+        stream: &mut StreamOf<I, E>,
+    ) -> PResult<I, O, E> {
+        let pre_state = stream.save();
+
+        #[allow(deprecated)]
+        let (mut a_errors, a_out) = debugger.invoke(&self.0, stream);
+
+        if a_out.is_ok() {
+            return (a_errors, a_out);
+        }
+
+        stream.revert(pre_state);
+
+        #[allow(deprecated)]
+        let (mut b_errors, b_out) = debugger.invoke(&self.1, stream);
+
+        if b_out.is_ok() {
+            a_errors.append(&mut b_errors);
```

Comment:
I guess the question of whether secondary errors from b should be preserved is an open one, but I'd err on the side of saying that they shouldn't. Not certain though.